### ISSUE TITLE
Matching Pre-releases

### ISSUE THREAD

**Issue by pinkforest** [rustsec/rustsec#690]
I went down the :rabbit: :hole: :carrot: on this

axum-core had the pattern:

```
    0.2.7 (and anything below) is vulnerable
    0.2.8 (and anything until 0.3.0.rc.1) is not vulnerable
    0.3.0-rc.1 is vulnerable
    0.3.0-rc.2 (and anything above) is not vulnerable
    0.3.0 - eventually will not be vulnerable
```

## Try 1 - Wild implicit arithmetric 1

_This has troubling implicit arithmetric as pre-releases arithmetric is a "wild guess" beyond semver part_

`patched = [">= 0.2.8, < 0.3.0", ">= 0.3.0-rc.2"]`

Results:

`Overlapping version ranges: [0.3.0-rc.2, 0.3.0-rc.2] and [0.2.8, 0.3.0) for key versions at line 8 column 1`

## Try 2 -Wild implicit arithmetric 2

_This also has troubling implicit arithmetric as pre-releases arithmetric is a "wild guess" beyond semver part_

```
patched = [">= 0.2.8, < 0.3.0", ">= 0.3.0"]
unaffected = ["= 0.3.0-rc.2"]
```

Results:

`Overlapping version ranges: [0.3.0-rc.2, 0.3.0-rc.2] and [0.2.8, 0.3.0) for key versions at line 8 column 1`

## Try 3 - a hack that "works"

So here we pin down "exact" release version for `= 0.3.0` and another for` >= 0.3.1`
```
patched = ["= 0.2.8", "= 0.3.0", ">= 0.3.1"]
unaffected = ["= 0.3.0-rc.2"]
```

This also works:
```
patched = ["= 0.2.8", "= 0.3.0-rc.2", "= 0.3.0", ">= 0.3.1"]
```

But the problem here is tooling e.g. cargo deny doesn't distinguish between patched `= 0.3.0` and it's pre-releases

cargo audit however would work fine: https://github.com/rustsec/advisory-db/pull/1417#discussion_r968323230

Also downstream would probably wonder why we have pinned patched like that.

## Question is.. 

What do we want to do here.. 

Do we want to take pre-releases into account in advisories e.g. whether we start to carry advisories against pre-releases especially if pre-release is vulnerable and the *real* release will not be ?

And if we were to address this considering all the tooling around how would we address it ?

The real proper fix to this would clarify and specify how pre-releases are taken into account which we don't currently explicitly define how the arithmetric catches them as pre-releases can be user-defined where there is no standard arithmetric.

**Comment by tarcieri** [comment]
Pre-release matching is a longstanding problem owing to SemVer's well-intentioned but in practice bizarre (and IMO broken) handling of prereleases.

The `semver` crate now implements lexicographic ordering of versions with `Ord` on `Version`, however anything that uses `VersionReq` directly will get the bizarre SemVer matching rules.

I've written at length about that here:

- https://github.com/dtolnay/semver/issues/172#issuecomment-848260770
- https://github.com/dtolnay/semver/issues/236

**Comment by pinkforest** [comment]
So our canonical thinking for this scenario that includes 0.3.0 as unaffected is:

```
patched = [">= 0.2.8, < 0.3.0-rc.1", ">= 0.3.0-rc.2"]
```

This seems to work in a way that 0.3.0 is not affected..

So our canon we can document is - in order:

```
0.3.0-rc.1
0.3.0.rc.2
0.3.0  <-- release after pre-releases is "greater" than pre-releases so the above pattern works to catch it in "patched"
```

**Comment by tarcieri** [comment]
@pinkforest can you break down which tools behave in what ways? The latter is wrong

**Comment by pinkforest** [comment]
I haven't found yet which one behaves the latter way - I am also confirming the lexicographic preference otherwise.

cargo audit works fine and cargo deny has general feature missing https://github.com/EmbarkStudios/cargo-deny/issues/455#issuecomment-1243859423 which can send a patch as soon as we have canonical expectation how it should be.

GHSA hasn't yet confirmed but asked in Zulip to confirm how the import / dependabot works to intepret it.

I think if we document the first as canonical with lexicographic sorting expectation then we could be fine / better I think in the long run to streamline it ?

**Comment by tarcieri** [comment]
FWIW the SemVer 2.0 spec defines the ordering as "precedence" in item 11 of:

https://github.com/semver/semver/blob/master/semver.md#semantic-versioning-specification-semver

**Comment by pinkforest** [comment]
So that "precedence" with pre-releases wasn't there before (?) or maybe the previous spec before 2.0 of that spec was ambiguous / broken or was it just the semver crate implementation was just broken .. I mean to me the spec - as of now - looks alright ?

The example on that precedence is pretty clear how it should work spec wise where < 1.0.0 trumps pre-releases:
`Example: 1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta < 1.0.0-beta < 1.0.0-beta.2 < 1.0.0-beta.11 < 1.0.0-rc.1 < 1.0.0.`

To me that spec (going by the first semver issue) `3.0.0-rc.0` should match `>= 2.0.0` going by above example and spec.

The "precedence" as described in 2.0 spec seems to make sense indeed and largely aligned what I would expect and something easy we can refer to and present as true canonical.

It also defines the build metadata with '+' e.g. openssl-src crates have it to denote the C-library version but unknown as to whether it should be used for matching.

Though should we make this work:

`patched = [">= 0.2.8, < 0.3", ">= 0.3.0-rc.2"]`

0.3 would denote "series" in general that includes the pre-releases - currently this does not work and `< 0.3` complains overlapping with `>= 0.3.0-rc.2`

**Comment by tarcieri** [comment]
Note that "precedence" is how versions are ordered and not how they necessarily satisfy requirements, particularly `semver::VersionReq`. The `rustsec` crate (now, thanks to @Shnatsel) matches requirements according to precedence, which is different from other tools like `cargo` (or anything else built on `semver::VersionReq`).

That said, I think the requirement you're after is:

```
patched = [">= 0.2.8, < 0.3.0-rc.1", ">= 0.3.0-rc.2"]
```

**Comment by pinkforest** [comment]
I wouldn't like to define / pin to 0.3.0-rc.1 when I could just say < 0.3 as whole series to apply the >= 0.2.8 patch series

e.g. can we make this work where `< 0.3`. matches everything `< 0.3.Y` including pre-releases

`patched = [">= 0.2.8, < 0.3", ">= 0.3.0-rc.2"]`

Current the above does not work and complains overlap when the patch part has not been defined in any way

**Comment by tarcieri** [comment]
> when I could just say < 0.3 as whole series to apply the >= 0.2.8 patch series

But the problem is `0.3.0-rc.2` is part of that range, hence the conflict.

**Comment by pinkforest** [comment]
I would entirely understand if I say `< 0.3.0` by spec but I'm wondering 

If it would be worthwhile to say `< 0.3` that does not have that last `.0` (patch) part 

This `0.3` could denote the whole  major/minor series as a whole including pre-releases on `0.3.0` ?

e.g. 
 - `0.3.0-rc.2 < 0.3.0` -> `true` 

but can

 - `0.3 < 0.3.0` -> `true` too ?

**Comment by tarcieri** [comment]
What you're proposing is similar to how `semver::VersionReq` works. Prereleases are excluded from matching unless specific explicitly.

Where this becomes problematic is that `> 0.3` will not match `0.4.0-pre`, or `0.5.0-pre`, etc per the SemVer 2.0 requirement matching rules.

We've managed to migrate away from `VersionReq`-style matching to something which uses precedence explicitly. My personal preference is to remain there versus reverting to `VersionReq`-style matching (which IMO does not make sense for our cases), and I also think it doesn't make sense to add `VersionReq`-style exceptions which are similar but not identical to the upstream SemVer 2.0 spec.

Admittedly it's not great for this particular case, but I also don't think this case has come up before. Really it's all a roundabout way of saying that `0.3.0-rc.1` is vulnerable and everything else `>= 0.2.8` (following precedence ordering) is not.

**Comment by pinkforest** [comment]
So yeah given all that what I learned today and given the inherent complexities -

Yeah I would also expect `> 0.3` to match `0.4.0-pre` . that doesn't seem like very logical to me either

I think requirement `patched = [">= 0.2.8, < 0.3.0-rc.1", ">= 0.3.0-rc.2"]` is indeed the proper and the tooling works properly to prevent any overlapping ranges conflicting which is really good for enforcing consistency - all I gotta do is clarify this somewhere in contributor doc / example if any at all if that is feasible.

Thanks :purple_heart: !
